### PR TITLE
apiserver: cancel context on timeout in WithTimeoutForNonLongRunningRequests

### PR DESCRIPTION
**Pull request by sttts**
Requests that block time out after 60sec, but keep the handler body running. This can lead to exhaustion of clients or other leaks. This PR adds a cancel func to the context of the request and calls it on timeout.

Note: we still do our own timeout handling as we don't trust the context to really cancel every blocking call we do.

This might explain why we see so many handler backtraces like https://gist.github.com/sttts/0ce972dc8a7911e4ca9eea7bf1ded5fa when an etcd node goes down with a hard poweroff. But it does not explain why we see oc to block for 15 minutes.

### CODE REVIEW

**Comment by sttts** [comment]
This also brings up the question whether we should cancel every context when the handler terminates, maybe after some extra time to allow possible bottom halfs of handlers to finish gracefully.

**Comment by sttts** [comment]
@deads2k anything that speaks against cancelling the context on timeout?

File: staging/src/k8s.io/apiserver/pkg/server/filters/timeout.go [hunk 1]
```diff
@@ -97,6 +106,7 @@ func (t *timeoutHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	case <-done:
 		return
 	case <-after:
+		cancelFunc()
```
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/server/filters/timeout.go]: Is there a technical difference between having people just include their `cancelFunc` logic inside of the `recordFn`?  Maybe change it's name, but appears logically equivalent from a code perspective.

Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/server/filters/timeout.go]: No, there isn't. The name `recordFn` is misleading. We should call it `onTimeoutFn` and cover both, metrics and cancellation.

Comment (sttts) [reply staging/src/k8s.io/apiserver/pkg/server/filters/timeout.go]: Fixed.

**Pull request merged**